vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/dataview/GridDataController.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2000-2020 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.vaadin.flow.component.grid.dataview;
+
+import java.util.ArrayList;
+import java.util.stream.Stream;
+
+import com.vaadin.flow.component.grid.Grid;
+import com.vaadin.flow.data.provider.DataCommunicator;
+import com.vaadin.flow.data.provider.DataController;
+import com.vaadin.flow.data.provider.DataKeyMapper;
+import com.vaadin.flow.data.provider.DataProvider;
+import com.vaadin.flow.data.provider.SizeChangeEvent;
+import com.vaadin.flow.data.provider.SizeChangeListener;
+import com.vaadin.flow.shared.Registration;
+
+/**
+ * DataController for Grid data communication.
+ *
+ * @param <T>
+ *         item type
+ * @since
+ */
+public class GridDataController<T> implements DataController<T> {
+
+    private Grid<T> grid;
+    private DataCommunicator<T> dataCommunicator;
+    private ArrayList<SizeChangeListener> listeners;
+
+    /**
+     * Constructs a new data controller for Grid.
+     *
+     * @param grid
+     *         Grid component for data controller.
+     */
+    public GridDataController(Grid<T> grid) {
+        this.grid = grid;
+        this.dataCommunicator = grid.getDataCommunicator();
+    }
+
+    /**
+     * Get the currently
+     *
+     * @return
+     */
+    public Stream<T> getCurrentItems() {
+        final DataKeyMapper<T> keyMapper = dataCommunicator.getKeyMapper();
+        return dataCommunicator.getActiveKeyOrdering().stream()
+                .map(keyMapper::get);
+    }
+
+    @Override
+    public DataProvider<T, ?> getDataProvider() {
+        return dataCommunicator.getDataProvider();
+    }
+
+    @Override
+    public Registration addSizeChangeListener(SizeChangeListener listener) {
+        if (listeners == null) {
+            listeners = new ArrayList<>(0);
+        }
+        listeners.add(listener);
+        return () -> listeners.remove(listener);
+    }
+
+    @Override
+    public int getDataSize() {
+        return dataCommunicator.getDataSize();
+    }
+
+    @Override
+    public Stream<T> getAllItems() {
+        return getDataProvider()
+                .fetch(dataCommunicator.buildQuery(0, Integer.MAX_VALUE));
+    }
+
+    /**
+     * Fire a size change event to all registered listeners.
+     */
+    public void fireSizeChangeEvent() {
```

Comment:
If someone does `dataProvider.refreshAll()` and the number of items has changed, then nothing will trigger this method.
Same goes for filtering changes through the data provider.

A clear indication that the size change listeners need to be in `DataConnector` or `AbstractSomethingSomething` ? So basically any component should get this stuff for free.

I'm not sure if we'll want to add this code now it will be removed in a few days anyway.